### ISSUE TITLE
[Feature Request]: Allow sorted `colors` prop with `ColorItem`

### ISSUE THREAD

**Issue by pm0u** [storybookjs/storybook#22067]
### Is your feature request related to a problem? Please describe

When using a typescale based on numbers (such as in tailwind), we want to have our object keys for colors be numbers:

```ts
...
  neutral: {
    00: "#FFFFFF",
    10: "#F5F6FF",
    20: "#D4D6EC",
    30: "#B1B3CE",
    40: "#878AAE",
    50: "#606384",
    60: "#464862",
    70: "#323346",
    80: "#1F202C",
    90: "#14151C",
  },
  ...
```

However, in typescript this throws a compiler error related to octals being used as object keys. So, a simple solution would be to make the `00` a string: `'00'` - however, this throws off the sorting order in the color palette:

<img width="484" alt="image" src="https://user-images.githubusercontent.com/410558/231612486-fbb7fab6-9de1-4200-a24f-0629870a112b.png">

Another option could be to use the suggested octal syntax:
<img width="742" alt="image" src="https://user-images.githubusercontent.com/410558/231612586-d6748ce4-2466-41fe-b1cd-916af2f4c631.png">

However this then presents the color as a single `0` when rendered (this is sort of a nit):

<img width="514" alt="image" src="https://user-images.githubusercontent.com/410558/231612648-fce84096-99f2-41f7-97f6-ffa924cd511e.png">

### Describe the solution you'd like

I would like to be able to control the sort order arbitrarily while providing a display name for a color.

### Describe alternatives you've considered

accepting that a single 0 is fine for the presentation of this color

### Are you able to assist to bring the feature to reality?

yes, I can

### Additional context

I have a branch here: https://github.com/storybookjs/storybook/compare/next...pm0u:storybook:sorted-color-palette?expand=1

that allows this feature. It's not the most robust implementation but if it's worth opening a PR i can. I'm running into several issues following the contributing guide here: https://storybook.js.org/docs/react/contribute/code

- multiple tests are failing on `yarn test`
- `yarn start` fails to start a sandbox

so I don't really know how to view my changes.

**Comment by lukeocodes** [comment]
+1 this

**Comment by shilman** [comment]
@Integrayshaun @JReinhold @kasperpeulen Can we come up with a workaround to achieve @pm0u 's desired outcome without any significant modifications to the `ColorPalette` block? I really don't want to make any changes (esp API changes) unless it's part of a bigger initiative to better support design tokens.

**Comment by JReinhold** [comment]
As you've probably also concluded, the only real solution is to use strings for keys, as JavaScript ignores leading zeros in numbers.

The wrong order is caused by `Object.keys` and `Object.values` sorting number-like keys.

A dirty workaround for now would be to add a space " " to each key, which from my experimentation stops the properties from being sorted:

```
  neutral: {
    "00 ": "#FFFFFF",
    "10 ": "#F5F6FF",
    "20 ": "#D4D6EC",
    "30 ": "#B1B3CE",
    ...
  },
```

I do think @pm0u approach in their branch where making colors an array instead of an object is the correct way, since it's the only way to ensure a stable order of properties.

I can come up with alternative solutions with other APIs that wouldn't be breaking, but they'd be worse.

**Comment by Integrayshaun** [comment]
I can't think of a better way to handle this without an array like @pm0u is suggesting. Looking back through my design token experiments, I didn't have a "000" case to reveal this problem.

![image](https://user-images.githubusercontent.com/18172605/232067280-71c7d12e-ac25-4b54-a542-d9fbf04357e7.png)

**Comment by pm0u** [comment]
For our use case, we have a number of scales similar to your screenshot. These are 10-90. The neutral scale includes 00 and 100 representing full white and black (technically, all scales would have these values at 00 and 100 in our system, but would be redundant).

It may be a somewhat limited use case. I had originally thought that this could be useful for languages that included diacritics as well, that could throw off the sort order. It seems like at least in chrome, the order is "preserved" _unless_ this is a parseable number (regardless if the key is a string or a number). In these cases, `0` behaves differently (at least in chrome)

```js
Object.keys({ 'Żółty': 1, 'Biały': 2, '00': 2, '300': 3, '100': 3, 'CHFJDHF': 200 })
> (6) ['100', '300', 'Żółty', 'Biały', '00', 'CHFJDHF']
```

**Comment by cdedreuille** [comment]
I agree with you on this one. To have better controls over the list of colours, an array of objects might be a better solution.
